### PR TITLE
ADBDEV-4936: Add assertions to return from dynamic_cast.

### PR DESCRIPTION
**Pull request by RekGRpth**
Add assertions to return from dynamic_cast.

dynamic_cast may return NULL, so I added assertions.

### CODE REVIEW

**Comment by KnightMurloc** [comment]
It seems to me that some of the fixes may not be necessary since the warnings are false positive.

When parsing DXL in ORCA, in most cases we expect that the elements will go in a strictly defined order and if this is not the case, an exception will be thrown.

For example OptimizerConfig:
We expect the first element to be EnumeratorConfig, so we put its ParseHandler at the top of the stack.
https://github.com/arenadata/gpdb/blob/c097366900c7699f6e10fe5f29a763792d34af41/src/backend/gporca/libnaucrates/src/parser/CParseHandlerOptimizerConfig.cpp#L153-L158
When parsing the next element, its startElement method will be called, which checks that this element is indeed EnumeratorConfig and otherwise throws an exception.
https://github.com/arenadata/gpdb/blob/c097366900c7699f6e10fe5f29a763792d34af41/src/backend/gporca/libnaucrates/src/parser/CParseHandlerEnumeratorConfig.cpp#L69-L77

As I understand it, problems can only be with optional elements, which can be parsed in any order, but in endElement it is expected in a strictly defined order. For example, Hint.

Do you think it's worth adding assert for required elements?

**Comment by RekGRpth** [comment]
> Do you think it's worth adding assert for required elements?

I think a debug GPOS_ASSERT won't hurt.

**Comment by KnightMurloc** [comment]
OK. Then I think the edits related to the parser should be made into a separate commit. It may also be worth do assert in a single line where possible. Something like:
```c
GPOS_ASSERT(pphEnumeratorConfig && pphStatisticsConfig && pphCTEConfig && pphDefoidsGPDB);
```

**Comment by RekGRpth** [comment]
> the edits related to the parser should be made into a separate commit

It so happened that the numbers associated with the parser are already spread across several tasks.

**Comment by KnightMurloc** [comment]
Yes, it won’t be possible to make edits to the parser in one commit (and maybe this is not a very good idea), but I think this patch should be divided into at least 2. And make edits to the parser separately.

**Comment by RekGRpth** [comment]
> make edits to the parser separately

https://github.com/arenadata/gpdb/pull/823

**Comment by RekGRpth** [comment]
> this patch should be divided into at least 2

done

File: src/backend/gporca/libgpopt/src/minidump/CMinidumperUtils.cpp [hunk 1]
```diff
@@ -213,7 +213,7 @@ CMinidumperUtils::Finalize(CMiniDumperDXL *pmdmp, BOOL fSerializeErrCtx)
 	CAutoTraceFlag atf3(EtraceSimulateNetError, false);
 	CAutoTraceFlag atf4(EtraceSimulateIOError, false);
 
-	if (NULL != CTask::Self() && fSerializeErrCtx)
```
Comment (KnightMurloc) [inline src/backend/gporca/libgpopt/src/minidump/CMinidumperUtils.cpp]: overwriting changes from another patch.

File: src/backend/gporca/libgpopt/src/minidump/CSerializableStackTrace.cpp [hunk 2]
```diff
@@ -60,7 +60,7 @@ CSerializableStackTrace::~CSerializableStackTrace()
 void
 CSerializableStackTrace::Serialize(COstream &oos)
 {
-	if (NULL == CTask::Self() || !CTask::Self()->HasPendingExceptions())
```
Comment (KnightMurloc) [inline src/backend/gporca/libgpopt/src/minidump/CSerializableStackTrace.cpp]: same

File: src/backend/gporca/libgpopt/src/operators/CExpressionHandle.cpp [hunk 3]
```diff
@@ -1374,9 +1374,7 @@ CExpressionHandle::PexprScalarRepChild(ULONG child_index) const
 		return pexprScalar;
 	}
 
-	GPOS_ASSERT(m_pexpr);
-
-	if (NULL != (*m_pexpr)[child_index]->Pgexpr())
```
Comment (KnightMurloc) [inline src/backend/gporca/libgpopt/src/operators/CExpressionHandle.cpp]: and in a number of other places.

File: src/backend/gporca/libgpopt/src/minidump/CMinidumperUtils.cpp (hunk 1 above)
Comment (RekGRpth) [reply src/backend/gporca/libgpopt/src/minidump/CMinidumperUtils.cpp]: indeed, fixed

File: src/backend/gporca/libgpopt/src/minidump/CSerializableStackTrace.cpp (hunk 2 above)
Comment (RekGRpth) [reply src/backend/gporca/libgpopt/src/minidump/CSerializableStackTrace.cpp]: indeed, fixed

File: src/backend/gporca/libgpopt/src/operators/CExpressionHandle.cpp (hunk 3 above)
Comment (RekGRpth) [reply src/backend/gporca/libgpopt/src/operators/CExpressionHandle.cpp]: indeed, fixed

**Review by KnightMurloc — APPROVED** [review]

**Pull request merged**